README.md
# Python Mattermost Driver (APIv4)

## Info

Currently WIP / Unstable

## Installation

`pip install mattermostdriver`

## Usage

```python
from mattermostdriver import Driver

foo = Driver({
    # Required options
    'url': 'mattermost.server.com',
    'login_id': 'user.name',
    'password': 'verySecret',
    # Optional / defaults to
	'scheme': 'https',
	'port': 8065,
	'basepath': '/api/v4',
	# Use False if self signed/insecure certificate
	'verify': True,
	# The interval the websocket will ping the server to keep the connection alive
	'timeout': 30,
})

foo.login() # Returns the response

foo.api.users.get_user_by_username('another.name') # Returns JSON

foo.api.channels.create_channel({
    'team_id': 'some_team_id',
    'name': 'awesome-channel',
    'display_name': 'awesome channel',
    'type': 'O'
})

foo.init_websocket(event_handler)
```

##### Available endpoints:
 - users
 - teams
 - channels
 - posts
 - files
 - preferences
 - system
 - webhooks
 - commands
 - compliance
 - cluster
 - brand
 - oauth
 - saml
 - ldap
 - jobs

See https://api.mattermost.com/v4/ to see which api requests are available.


CHANGELOG.rst
Unreleased
'''''''''''

Added
- HTTP proxy support for http + websocket calls
  - Websocket calls are now using aiohttp

7.3.2
''''''
Fixed
 - Login failed in version 6.3.0 #112

Thanks a lot to @iprok for their contribution!

7.3.1
''''''
Fixed
 - Missing documentation endpoints #95
 - Incorrect submit dialog #94
 - Disconnect not working #91

Requirement changes
 - websockets bumped from 8.1 to 9.1

Thanks a lot to @unode and @wonsungkang for their contribution!

7.3.0
''''''
Added
 - New keepalive option, `keepalive` and `keepalive_delay`, default `False` #86
 - New `websocket_kw_args` option which is passed to the websocket library on `connect()` #86

Thanks a lot to @ttuffin and @attzonko for their contribution!

7.2.0
''''''
Requirement changes
 - websocket bumped from 6.0 to 8.1
 - requests bumped from 2.20.0 to 2.25.1

Added
 - Add Driver.disconnect function #80
 - Raise errors from None to reduce traceback noise #69

Fixes
 - Correct url for role by name #78
 - Add oldest unread endpoint #76

Documentation:
 - Improve documentation about verify #64

Thanks a lot to @jneeven for his contribution!

7.1.0
'''''
Added endpoints
 - `IntegrationActions`

Thanks a lot to @ritelle for his contribution!

7.0.1
'''''
Fixes
 - Log `WebSocket authentication failed` even if WebSocket authentication was successful

Thanks a lot to @icedevml for his contribution!

7.0.0
'''''
Added Endpoints:
 - `opengraph`
 - `bots`

Fixes:
 - Some pylint rules that failed have been fixed or where suppressed

Added:
 - Support for Python 3.8
 - Github actions with pylint is now active

BREAKING CHANGES:
 - Support for python 3.4 was dropped

Thanks a lot to @maxbrunet for his contribution!

6.3.1
'''''
Fixes
 - Authentication without `auth` option works again

6.3.0
'''''
Features:
 - Added authentication with .netrc

Thanks a lot to @apfeiffer1 for his contribution!

6.2.0
'''''
Fixes:
 - Fix documentation to make Json valid

Features:
 - Added a configurable timeout #40

Thanks a lot to @sahasrara62 for his contribution!

6.1.3
'''''
Fixes:
 - Add missing endpoints to the driver

Thanks a lot to @opalmer for this fix!

6.1.2
'''''
Fixes:
 - Don't parse non JSON errors #31
 - Logout before cleaning up #37

Thanks a lot to @maxbrunet for these two fixes!

6.1.1
'''''
Update requirements:
- requests@2.20.0

6.1.0
'''''
Added Endpoints:
- Reactions

Thanks a lot to @aedho3yn for this!

6.0.1
'''''
Fixes:
 - Unable to create custom emojies #33
 - Use ResourceNotFound for 404 #34

Thanks a lot to @Sakarah and @maxbrunet for these fixes!

6.0.0
'''''
POSSIBLE BREAKING CHANGES:
 Requirements have been updated:
 - websockets==6.0

This requirement (and requests from 5.0.0) are now included in the setup.py file.
If you didn't need to update before it is possible you do now,
which could cause a breaking change this time.
Not exactly sure about this, but better safe then sorry!

5.0.1
'''''
Fixes:
 - Do not try to parse responses as json when it's clear they are not. Thanks Kenan!

5.0.0
'''''
POSSIBLE BREAKING CHANGES:
 Requirements have been updated:
 - requests==2.19.1
 - websockets==5.0

Fixes:
 - The error message 'Websocket authentication failed' was not always correct

4.6.0
'''''
Release for Mattermost 5.0 api changes.

Yes, I skipped the 4.9 release because of holiday time ;-)

Added endpoints:
 - `scheme` as a whole new endpoint
 - `update_scheme_derived_roles_of_channel_member` in `channels`
 - `set_channel_scheme` in `channels`
 - `convert_channel` in `channels`
 - `update_scheme_derived_roles_of_team_member` in `teams`
 - `delete_team_icon` in `teams`
 - `set_team_scheme` in `teams`
 - `get_stats` in `users`

Fixes:
 - `create_user` in `users` was missing the query parameters

4.5.0
'''''
Release for Mattermost 4.8 api changes.

Yes this release is quite a bit behind schedule, sorry for that!

Added endpoints:
 - `status` as a whole new endpoint
 - `roles` as a whole new endpoint
 - `test_aws_s3_connection` in `system`
 - `get_configuration_environment` in `system`
 - `send_test_email` in `system`
 - `create_ephemeral_post` in `posts`
 - `get_team_icon` in `teams`
 - `set_team_icon` in `teams`

Fixes:
 - Example for `call_webhook` was incorrect #26
 - A slash was missing in `update_user_authentication_method` #27

4.4.0
'''''
Release for Mattermost 4.7 api changes.

Added endpoints:
 - `get_custom_emoji_by_name` in `emoji`
 - `search_custom_emoji` in `emoji`
 - `autocomplete_custom_emoji` in `emoji`
 - `autocomplete_channels` in `channels`

Fixes:
 - `update_user_authentication_method` was missing the request body

4.3.2
'''''
Fixes
 - https://github.com/Vaelor/python-mattermost-driver/issues/24

4.3.1
'''''
Added endpoints
 - `create_user_access_token` in `/users`

4.2.1
'''''
Fixes
 - https://github.com/Vaelor/python-mattermost-driver/pull/21
 - https://github.com/Vaelor/python-mattermost-driver/pull/22

Thanks to @dan-klasson for these!!

4.2.0
'''''
Release for Mattermost 4.6 api changes.

Added endpoints:
 - `get_user_access_token` in `/users`
 - `search_tokens` in `/users`
 - `update_user_authentication_method` in `/users`

4.1.0
'''''
This release mostly improves on the documentation.

The sphinx theme has been changed to the readthedocs one.

This also adds a `debug` option, which enables a very verbose log output.
Be careful, as everything, even your mattermost password when you log in,
is readable in the log output!
This is definitely not for production usage!

4.0.2
'''''
This release makes some internal changes on how the endpoints are accessed.

Since this works much better then using `api['endpoint']` has been deprecated for the next Major release.

Fixes https://github.com/Vaelor/python-mattermost-driver/issues/5


4.0.1
'''''
The release 4.0.0 was not quite correct, since the following changes did not really happen, only the api documentation for mattermost 4.4.0 changed.

.. code::


src/driver.py
import json

import asyncio
import requests
import logging
import time

from src.client import Client
from src.endpoints.brand import Brand
from src.endpoints.channels import Channels
from src.endpoints.cluster import Cluster
from src.endpoints.commands import Commands
from src.endpoints.compliance import Compliance
from src.endpoints.files import Files
from src.endpoints.jobs import Jobs
from src.endpoints.ldap import LDAP
from src.endpoints.oauth import OAuth
from src.endpoints.posts import Posts
from src.endpoints.preferences import Preferences
from src.endpoints.saml import SAML
from src.endpoints.system import System
from src.endpoints.teams import Teams
from src.endpoints.users import Users
from src.endpoints.webhooks import Webhooks
from src.websocket import Websocket

logging.basicConfig(level=logging.INFO)
log = logging.getLogger('mattermost.api')


class Driver:

	defaultOptions = {
		'scheme': 'https',
		'url': 'localhost',
		'port': 8065,
		'basepath': '/api/v4',
		'verify': True,
		'timeout': 30,
		'login_id': None,
		'password': None,
	}

	def __init__(self, options):
		self.options = self.defaultOptions.copy()
		self.options.update(options)
		self.driver = self.options
		self.client = Client(self.options)
		self.routes = {}
		self.websocket = None

	def init_websocket(self, event_handler):
		self.websocket = Websocket(self.options, self.client.token)
		loop = asyncio.get_event_loop()
		loop.run_until_complete(self.websocket.connect(event_handler))
		return loop

	def login(self):
		result = self.users_api().login_user({
			'login_id': self.options['login_id'],
			'password': self.options['password'],
		})
		if result.status_code == 200:
			self.client.token = result.headers['Token']
		return result

	def logout(self):
		self.users_api().logout_user()

	def users_api(self):
		if 'users' not in self.routes:
			self.routes['users'] = Users(self.client)
		return self.routes['users']

	def teams_api(self):
		if 'teams' not in self.routes:
			self.routes['teams'] = Teams(self.client)
		return self.routes['teams']

	def channels_api(self):
		if 'channels' not in self.routes:
			self.routes['channels'] = Channels(self.client)
		return self.routes['channels']

	def posts_api(self):
		if 'posts' not in self.routes:
			self.routes['posts'] = Posts(self.client)
		return self.routes['posts']

	def files_api(self):
		if 'files' not in self.routes:
			self.routes['files'] = Files(self.client)
		return self.routes['files']

	def preferences_api(self):
		if 'preferences' not in self.routes:
			self.routes['preferences'] = Preferences(self.client)
		return self.routes['preferences']

	def webhooks_api(self):
		if 'webhooks' not in self.routes:
			self.routes['webhooks'] = Webhooks(self.client)
		return self.routes['webhooks']

	def commands_api(self):
		if 'commands' not in self.routes:
			self.routes['commands'] = Commands(self.client)
		return self.routes['commands']

	def system_api(self):
		if 'system' not in self.routes:
			self.routes['system'] = System(self.client)
		return self.routes['system']

	def compliance_api(self):
		if 'compliance' not in self.routes:
			self.routes['compliance'] = Compliance(self.client)
		return self.routes['compliance']

	def cluster_api(self):
		if 'cluster' not in self.routes:
			self.routes['cluster'] = Cluster(self.client)
		return self.routes['cluster']

	def brand_api(self):
		if 'brand' not in self.routes:
			self.routes['brand'] = Brand(self.client)
		return self.routes['brand']

	def oauth_api(self):
		if 'oauth' not in self.routes:
			self.routes['oauth'] = OAuth(self.client)
		return self.routes['oauth']

	def saml_api(self):
		if 'saml' not in self.routes:
			self.routes['saml'] = SAML(self.client)
		return self.routes['saml']

	def ldap_api(self):
		if 'ldap' not in self.routes:
			self.routes['ldap'] = LDAP(self.client)
		return self.routes['ldap']

	def jobs_api(self):
		if 'jobs' not in self.routes:
			self.routes['jobs'] = Jobs(self.client)
		return self.routes['jobs']
